### PR TITLE
Resource modules api and Testing

### PR DESCRIPTION
**Pull request by that1guy15**
New resource modules:

- aos_vni_pool
- aos_ip_pool

Testing now includes ansible playbook for integration test. 
aos.py moved to library for cleaner install and use.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,16 +2,37 @@
     
 # AOS Ansible Modules
 
-This project is under active development. 
-Documentation and example code is coming soon, so please stay tuned!
+Ansible modules used to interact with Apstra AOS.
 
-Questions? Comments? Please use the github issues list.
+## Current Modules
```
Comment (roman-verchikov) [inline README.md]: Not sure about the value of this section given that it simply enumerates existing files and doesn't provide any documentation on how to use it.

File: README.md [hunk 2]
```diff
@@ -2,16 +2,37 @@
...
+* aos_vni_pool
+* aos_ip_pool
+
+### Blueprint
+* aos_blueprint_deploy
+
+## Installation
+Clone repo into Library directory of your Ansible project
 
+`cd ./library
```
Comment (roman-verchikov) [inline README.md]: this will render into single line code statement. Use multi-line code syntax instead (```)

File: README.md [hunk 3]
```diff
@@ -2,16 +2,37 @@
...
+git clone https://github.com/Apstra/aos-ansible.git`
 
-# Testing
+Install aos-ansible requrements
 
-`nosetests -v`
+`cd aos-ansible
```
Comment (roman-verchikov) [inline README.md]: same as above

File: README.md [hunk 4]
```diff
@@ -2,16 +2,37 @@
...
+* aos_ip_pool
+
+### Blueprint
+* aos_blueprint_deploy
+
+## Installation
+Clone repo into Library directory of your Ansible project
 
+`cd ./library
+git clone https://github.com/Apstra/aos-ansible.git`
```
Comment (roman-verchikov) [inline README.md]: it might be better to use git submodules in this case, which would allow to reference `aos-ansible` repo instead of keeping a copy.

File: README.md [hunk 5]
```diff
@@ -2,16 +2,37 @@
...
 
-`nosetests -v`
+`cd aos-ansible
+pip install -r requirements.txt`
 
+Add library reference to ansible.cfg if not in standard project library
+
+
+## Contribution
+[Adding new modules or roles](https://github.com/Apstra/aos-ansible/blob/develop/CONTRIBUTING.md)
```
Comment (roman-verchikov) [inline README.md]: I guess it would be better to simply state

> See CONTRIBUTING.md

The thing about HTTP links is that they tend to break. In this particular example the most likely case would be if we ditched `develop` branch and only used `master`. Other examples include moving source code to a different repository, renaming file, etc.

File: README.md [hunk 6]
```diff
@@ -2,16 +2,37 @@
...
+
+# License
+Apache 2.0
```
Comment (roman-verchikov) [inline README.md:29]: License need to be revisited

File: inventory.ini [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[infra]
+# https://cloud.ravellosystems.com/#/0/apps/3125678020704/vms/?selectedIds=;3343793684176704512
+aos3.0-l2-evpn	ansible_host=aos-testaosansiblel2ev-rh9yt9ie.srv.ravcloud.com
```
Comment (roman-verchikov) [inline inventory.ini]: I guess it would be ok as first implementation, but later it should be replaced with travis encrypted environment variables: https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings

File: library/aos_ip_pool.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
+#!/usr/bin/python
+#
+# (c) 2017 Apstra Inc, <community@apstra.com>
+#
+# This file is part of Ansible
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py]: I don't think this statement is accurate, given that this module is not going to be core ansible module.

Also it contradicts Ansible requirements for license headers: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright-and-license.

> The license declaration should be ONLY one line, not the full GPL prefix.:

File: library/aos_ip_pool.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+def validate_subnets(module, subnets, addr_type):
+    """
+    Validate IP subnets provided are valid and properly formatted
+    :param module: Ansible built in
+    :param subnets: list
+    :param addr_type: int
+    :return: bool
+    """
+    for i, subnet in enumerate(subnets, 1):
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py]: `i` is unused

File: library/aos_ip_pool.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        try:
+            results = ipaddress.ip_network(subnet)
+            if results.version != addr_type:
+                module.fail_json(msg="Subnet {} does not match ip_version {} "
+                                     "configured for the pool "
+                                 .format(subnet, addr_type))
+
+        except ValueError as e:
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py]: there can also be a `TypeError` in case user specifies `null` as subnet value

File: library/aos_ip_pool.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+                             name=new_pool['display_name'],
+                             id=resp['id'],
+                             value=new_pool)
+
+        module.exit_json(changed=False,
+                         name=new_pool['display_name'],
+                         id={},
+                         value={})
+
+    else:
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py:206]: `elif my_pool['subnets']:`

File: library/aos_vni_pool.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+      - An existing AOS session as obtained by M(aos_login) module.
+    required: true
+  name:
+    description:
+      - Name of the VNI Pool to manage.
+        Only one of I(name), I(id) or I(content) can be set.
+  id:
+    description:
+      - AOS Id of the VNI Pool to manage.
+        Only one of I(name), I(id) or I(content) can be set.
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: I don't see `content` defined anywhere

File: library/aos_vni_pool.py [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+    It's using the AOS REST API.
+options:
+  session:
+    description:
+      - An existing AOS session as obtained by M(aos_login) module.
+    required: true
+  name:
+    description:
+      - Name of the VNI Pool to manage.
+        Only one of I(name), I(id) or I(content) can be set.
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: `content` is referenced in documentation, but is not used anywhere

File: library/aos_ip_pool.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    :param session: dict
+    :param endpoint: str
+    :param my_pool: dict
+    :return: dict
+    """
+    margs = module.params
+
+    if not my_pool:
+
+        if 'name' not in margs.keys():
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py:192]: `'name'` is also required for update

File: library/aos_ip_pool.py [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+            module.fail_json(msg="name is required to create a new resource")
+
+        new_pool = {"subnets": get_subnets(margs['subnets']),
+                    "display_name": margs['name'],
+                    "id": margs['name']}
+
+        if not module.check_mode:
+            resp = aos_post(session, endpoint, new_pool)
+
+            module.exit_json(changed=True,
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py]: this is repeated below

File: library/aos_ip_pool.py [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    else:
+        endpoint = V4_ENDPOINT
+
+    my_pool = find_resource_item(margs['session'],
+                                 endpoint,
+                                 resource_name=item_name,
+                                 resource_id=item_id)
+
+    if margs['state'] == 'absent':
+        ip_pool_absent(module, margs['session'], endpoint, my_pool)
```
Comment (roman-verchikov) [inline library/aos_ip_pool.py]: the contract between caller and callee and variable naming is too obscure. What is `my_pool`? Why is it `my`? Who's "me"?

Comment (roman-verchikov) [reply library/aos_ip_pool.py]: I'd suggest to rewrite using clear contract, something along these lines:

```python
target_state = margs["state"]

pool = find_resource_item(...)

if target_state == "absent":
    aos_pool_delete_by_id(session, pool["id"])
elif target_state == "present":
    pool_already_exists = pool != {}
    if pool_already_exists:
        aos_pool_update(session, pool["id"], pool)
    else:
        aos_pool_create(session, new_pool)
```

File: library/aos_vni_pool.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+  sample: {'...'}
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from library.aos import aos_post, aos_put, aos_delete, find_resource_item
+
+ENDPOINT = 'resources/vni-pools'
+
+
+def validate_ranges(module, ranges):
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: identical function already present in `aos_asn_pool`. Isolate and reuse

File: library/aos_vni_pool.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+                                 "integer instead of {} "
+                             .format(i, type(vni_range[1])))
+        elif vni_range[1] <= vni_range[0]:
+            module.fail_json(msg="2nd element of vni_range {} must be "
+                                 "bigger than 1st ".format(i))
+        elif vni_range[0] <= 4095 or vni_range[1] >= 16777213:
+            module.fail_json(msg="Range {} must be a valid range between 4096"
+                                 " and 16777214"
+                             .format(vni_range[0]))
+        i += 1
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: one of the reasons you should isolate functions and reuse them: i+=1 is not needed here, yet copy/pasted from old implementation, which spreads bad code further

File: library/aos_vni_pool.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+    if not module.check_mode:
+        aos_delete(session, ENDPOINT, my_pool['id'])
+
+    module.exit_json(changed=True,
+                     name=my_pool['display_name'],
+                     id=my_pool['id'],
+                     value={})
+
+
+def vni_pool_present(module, session, my_pool):
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py:165]: the structure is identical to `ip_pool_present` and other `*_present` functions in other modules with all the mistakes copy/pasted. Isolate and reuse.

File: tests/roles/test_clean_up_tests/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+      ip_version: 6
+    register: ip6pool
+
+  - name: test_delete_vni_pool
+    local_action:
+      module: aos_vni_pool
+      session: "{{ aos_session }}"
+      id: "my-vni-pool"
+      state: absent
+    register: vnipool
```
Comment (roman-verchikov) [inline tests/roles/test_clean_up_tests/tasks/main.yml]: newline

File: test_playbook.yaml [hunk 21]
```diff
@@ -0,0 +1,25 @@
+---
```
Comment (roman-verchikov) [inline test_playbook.yaml:1]: Might also be good to run this test in `checked` mode to do pure unit testing.

File: test_playbook.yaml [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+        - include_role:
+            name: test_pb_aos_ip_pool
+        - include_role:
+            name: test_pb_aos_blueprint_deploy
+      rescue:
+        - include_role:
+            name: test_clean_up_tests
+        - name: Failure detected in test roles
+          fail:
+            msg: "Failure detected in test roles, cleaning up and reset environment"
```
Comment (roman-verchikov) [inline test_playbook.yaml]: add newline

File: tests/roles/test_pb_aos_asn_pool/tasks/main.yml [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        - asnpool|changed
+        - asnpool.id == "my-asn-pool"
+        - asnpool.name == "my-asn-pool"
+        - asnpool.value.ranges == [asn_pool1]
+
+  - name: test_update_asn_pool_by_name
+    local_action:
+      module: aos_asn_pool
+      session: "{{ aos_session }}"
+      name: "{{ asnpool.name }}"
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_asn_pool/tasks/main.yml:130]: test case when both name and id are provided

File: tests/roles/test_pb_aos_asn_pool/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+    register: asnpool
+
+  - name: check_test_delete_asn_pool
+    assert:
+      that:
+        - asnpool|success
+        - asnpool|changed
+        - asnpool.id == "my-asn-pool"
+        - asnpool.name == "my-asn-pool"
+        - asnpool.value == {}
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_asn_pool/tasks/main.yml]: newline

File: tests/roles/test_pb_aos_blueprint_deploy/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+      module: aos_blueprint_deploy
+      session: "{{ aos_session }}"
+      name: "vpod-evpn"
+    register: bp_deploy
+
+  - name: check_test_deploy_vpod-evpn_blueprint
+    assert:
+      that:
+        - bp_deploy|success
+        - bp_deploy|changed == false
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_blueprint_deploy/tasks/main.yml]: newline

File: tests/roles/test_pb_aos_ip_pool/tasks/main.yml [hunk 26]
```diff
@@ -0,0 +1,249 @@
+
+  - set_fact:
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_ip_pool/tasks/main.yml:2]: why not simply use list of pools?

```yaml
- set_fact:
    ip_pools:
        - network: '192.168.59.0/24'
        - network: '192.168.60.0/24'
        - network: '10.10.10.0/23'
        - network: '192.168.70.0/24'
        - network: '10.10.20.0/23'
```

File: tests/roles/test_pb_aos_ip_pool/tasks/main.yml [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+        network: '192.168.60.0/24'
+      ip_pool3:
+        network: '10.10.10.0/23'
+      ip_pool4:
+        network: '192.168.70.0/24'
+      ip_pool5:
+        network: '10.10.20.0/23'
+
+  - set_fact:
+      ipv6_pool1:
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_ip_pool/tasks/main.yml:15]: same as above

File: tests/roles/test_pb_aos_ip_pool/tasks/main.yml [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+      port: 443
+    run_once: true
+
+  - name: test_create_new_ipv4_pool_invalid_subnet
+    local_action:
+      module: aos_ip_pool
+      session: "{{ aos_session }}"
+      name: "my-ip-pool"
+      subnets:
+        - '192.168.359.0/24'
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_ip_pool/tasks/main.yml:41]: also test for `null` and empty input

File: tests/roles/test_pb_aos_ip_pool/tasks/main.yml [hunk 29]
```diff
@@ -0,0 +1,249 @@
...
+    register: ip6pool
+
+  - name: check_test_delete_ipv6_pool
+    assert:
+      that:
+        - ip6pool|success
+        - ip6pool|changed
+        - ip6pool.id == "my-ipv6-pool"
+        - ip6pool.name == "my-ipv6-pool"
+        - ip6pool.value == {}
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_ip_pool/tasks/main.yml]: newline

File: tests/roles/test_pb_aos_login/tasks/main.yml [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+    ignore_errors: True
+
+  - debug:
+      msg: "{{ test.msg }}"
+
+  - name: check_test_aos_login_incorrect
+    assert:
+      that:
+        - test|failed
+        - "'Invalid credentials' in test.msg"
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_login/tasks/main.yml]: newline

File: tests/roles/test_pb_aos_vni_pool/tasks/main.yml [hunk 31]
```diff
@@ -0,0 +1,119 @@
...
+    register: vnipool
+
+  - name: check_test_delete_vni_pool
+    assert:
+      that:
+        - vnipool|success
+        - vnipool|changed
+        - vnipool.id == "my-vni-pool"
+        - vnipool.name == "my-vni-pool"
+        - vnipool.value == {}
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_vni_pool/tasks/main.yml]: newline

File: library/aos_vni_pool.py [hunk 32]
```diff
@@ -0,0 +1,306 @@
...
+                                 id=my_pool['id'],
+                                 value=new_pool)
+
+            module.exit_json(changed=False,
+                             name=new_pool['display_name'],
+                             id={},
+                             value={})
+
+
+def vni_pool(module):
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py:214]: all the `*_pool` functions share the same pattern:

1. validate input
2. check if resource exists
3. if target state is "absent" - delete resource
4. if target state is "present" and resource does not exist yet - create it
5. if target state is "present" and resource already exists - update existing resource

All of this can be isolated and reused with minor changes to avoid copy/pasting

File: library/aos_vni_pool.py [hunk 33]
```diff
@@ -0,0 +1,306 @@
...
+    """
+    Main function to create, change or delete AOS VNI resource pool
+    """
+    margs = module.params
+
+    item_name = False
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: `item_name` and `item_id` must either be string or `None`

**Review by roman-verchikov — CHANGES_REQUESTED** [review]
Way too much copy/pasting. All the coding mistakes are spreading across codebase very quickly with this approach.

All the license headers should be revisited and made single line statements according to Ansible module development guide.

File: tox.ini [hunk 34]
```diff
@@ -6,12 +6,20 @@ envlist = py37, flake8
 deps =
     -rrequirements.txt
     -rtest-requirements.txt
+passenv =
+    VAULT_PASSWORD
 commands =
+    touch vault_pass.txt
+    bash -ec 'if [[ $VAULT_PASSWORD ]]; then echo $VAULT_PASSWORD >> vault_pass.txt; fi'
```
Comment (roman-verchikov) [inline tox.ini]: I guess you need to overwrite instead of appending

```
echo $VAULT_PASSWORD > vault_pass.txt
```

Comment (roman-verchikov) [reply tox.ini]: also password may contain whitespaces, so it would be better to use quotes:

```
if [[ ${VAULT_PASSWORD} ]]; then echo "${VAULT_PASSWORD}" > vault_pass.txt; fi
```

File: tox.ini [hunk 35]
```diff
@@ -6,12 +6,20 @@ envlist = py37, flake8
 deps =
     -rrequirements.txt
     -rtest-requirements.txt
+passenv =
+    VAULT_PASSWORD
 commands =
+    touch vault_pass.txt
```
Comment (roman-verchikov) [inline tox.ini]: I think touch is not necessary, if `${VAULT_PASSWORD}` is set, then redirecting (`echo ${VAULT_PASSWORD}`) will create new file if it doesn't exist, and if `VAULT_PASSWORD` is not set, then you don't even need a vault_pass.txt.

**Review by roman-verchikov — CHANGES_REQUESTED** [review]

File: inventory.ini [hunk 36]
```diff
@@ -0,0 +1,18 @@
+$ANSIBLE_VAULT;1.1;AES256
```
Comment (roman-verchikov) [inline inventory.ini:1]: thats cool!

File: tox.ini [hunk 37]
```diff
@@ -6,12 +6,20 @@ envlist = py37, flake8
...
+    bash -ec 'if [[ $VAULT_PASSWORD ]]; then echo $VAULT_PASSWORD >> vault_pass.txt; fi'
     nosetests -v
+    ansible-playbook test_playbook.yaml --vault-password-file vault_pass.txt
+whitelist_externals =
+    /bin/bash
+    /usr/bin/touch
```
Comment (roman-verchikov) [inline tox.ini]: remove `touch`

File: TESTING.md [hunk 38]
```diff
@@ -0,0 +1,29 @@
...
+
+#### Run test (tox)
+
+NOTE: Ansible playbook `test_playbook.yaml` relies on the AOS instance
+from `ansible.cfg` to be running or these tests will fail
+
+~~~~
+pip install -r requirements.txt
+pip install -r test-requirements.txt
+tox
```
Comment (roman-verchikov) [inline TESTING.md:25]: tox should take care of dependencies. The only requirement to run it - is to have tox installed, so the only thing a user need to do to run tests is to run tox:

```
tox
```

File: TESTING.md [hunk 39]
```diff
@@ -0,0 +1,29 @@
...
+export VAULT_PASSWORD=<password for inventory.ini>
+~~~
+
+Alternativley you can replace `inventory.ini` with an unencrypted
+inventory of your own AOS test server.
+
+#### Run test (tox)
+
+NOTE: Ansible playbook `test_playbook.yaml` relies on the AOS instance
+from `ansible.cfg` to be running or these tests will fail
```
Comment (roman-verchikov) [inline TESTING.md]: you mean in inventory.ini?

File: library/aos_ip_pool.py (hunk 9 above)
Comment (roman-verchikov) [reply library/aos_ip_pool.py]: not addressed: remove `i` and `enumerate`

File: library/aos_vni_pool.py [hunk 40]
```diff
@@ -0,0 +1,295 @@
...
+
+def validate_ranges(ranges):
+    """
+    Validate VNI ranges provided are valid and properly formatted
+    :param ranges: list
+    :return: bool
+    """
+    errors = []
+
+    for i, vni_range in enumerate(ranges, 1):
```
Comment (roman-verchikov) [inline library/aos_vni_pool.py]: remove `i` and `enumerate`, they are unused

File: tests/roles/test_clean_up_tests/tasks/main.yml [hunk 41]
```diff
@@ -0,0 +1,54 @@
...
+    local_action:
+      module: aos_ip_pool
+      session: "{{ aos_session }}"
+      id: "{{item}}"
+      state: absent
+      ip_version: ipv6
+    register: ip6pool
+    with_items:
+          - "my-ipv6-pool"
+          - "my-blank-ipv6-pool"
```
Comment (roman-verchikov) [inline tests/roles/test_clean_up_tests/tasks/main.yml]: missing newline (and this nasty red symbol!)

File: test_playbook.yaml [hunk 42]
```diff
@@ -0,0 +1,34 @@
...
+            name: test_clean_up_tests
+        - name: Failure detected in test roles
+          fail:
+            msg: "Failure: {{ansible_failed_task.name}}: {{ansible_failed_result}}"
+      always:
+        - include_role:
+            name: test_clean_up_tests
+        - name: Cleaning up test roles and resetting test enviroment
+          debug:
+            msg: "Cleaning up after the mess. Nothing to see here, move along."
```
Comment (roman-verchikov) [inline test_playbook.yaml]: missing newline

File: tests/roles/test_pb_aos_blueprint_deploy/tasks/main.yml [hunk 43]
```diff
@@ -0,0 +1,26 @@
...
+      module: aos_blueprint_deploy
+      session: "{{ aos_session }}"
+      name: "vpod-evpn"
+    register: bp_deploy
+
+  - name: check_test_deploy_vpod-evpn_blueprint
+    assert:
+      that:
+        - bp_deploy|success
+        # - bp_deploy|changed == false
```
Comment (roman-verchikov) [inline tests/roles/test_pb_aos_blueprint_deploy/tasks/main.yml]: remove commented line

File: requirements.txt [hunk 44]
```diff
@@ -1,4 +1,3 @@
-ansible>=2.7.9
+ansible==2.7.9
 requests==2.21.0
-tox==3.7.0
-urllib3==1.24.1
\ No newline at end of file
+tox==3.7.0
```
Comment (roman-verchikov) [inline requirements.txt:3]: `tox` is not a dependency for any of the ansible modules, and should be removed from `requirements.txt`

File: test-requirements.txt [hunk 45]
```diff
@@ -2,4 +2,5 @@ flake8==3.7.7
 mock==2.0.0
 nose==1.3.7
 pylint==2.3.1
-tox==3.7.0
\ No newline at end of file
+pytest==4.5.0
+tox==3.7.0
```
Comment (roman-verchikov) [inline test-requirements.txt]: Even in this case, `tox` is not a dependency of tests (`tox` module is not imported by any test), instead it's an external tool used to run the tests. I guess the only place where `tox` should be mentioned is README and perhaps CI test script, where tox must be installed.

File: test-requirements.txt [hunk 46]
```diff
@@ -2,4 +2,5 @@ flake8==3.7.7
 mock==2.0.0
 nose==1.3.7
```
Comment (roman-verchikov) [inline test-requirements.txt]: nose dependency should be removed if you decided to switch to pytest, keeping multiple test libraries is confusing

File: tests/test_aos_login.py [hunk 47]
```diff
@@ -1,5 +1,4 @@
 from ansible.compat.tests.mock import patch
-from nose.tools import assert_equals
 import library.aos_login as aos_login
```
Comment (roman-verchikov) [inline tests/test_aos_login.py:6]: `AnsibleExitJson` and `AnsibleFailJson` classes are unused

File: tests/test_aos_login.py [hunk 48]
```diff
@@ -41,4 +40,4 @@ def test_aos_login_success(mock_module, mock_aos_login):
 
     mock_aos_login.return_value = mock_session
     resp = aos_login.aos_login()
```
Comment (roman-verchikov) [inline tests/test_aos_login.py]: aos_login() expects one argument which is not provided. Not sure how this works

File: tests/test_aos_vni_pool.py [hunk 49]
```diff
@@ -0,0 +1,107 @@
...
+@patch('library.aos_vni_pool.AnsibleModule')
+def test_aos_vni_pool_success(mock_module, mock_aos_vni_pool):
+    """
+    aos_vni_pool - test module returned data
+    """
+    mock_session = {"server": "foo",
+                    "token": "bar"}
+
+    mock_aos_vni_pool.return_value = mock_session
+    resp = aos_vni_pool.vni_pool()
```
Comment (roman-verchikov) [inline tests/test_aos_vni_pool.py]: vni_pool() requires module passed as argument. I guess `aos_vni_pool` is mocked, and it doesn't really test anything?.. Not sure how it works

File: tests/test_aos_vni_pool.py [hunk 50]
```diff
@@ -0,0 +1,107 @@
...
+                       'ranges': {'required': False,
+                                  'type': 'list',
+                                  'default': []}},
+        mutually_exclusive=[('name', 'id')],
+        required_one_of=[('name', 'id')], supports_check_mode=True)
+
+
+@patch('library.aos_vni_pool.vni_pool')
+@patch('library.aos_vni_pool.AnsibleModule')
+def test_aos_vni_pool_success(mock_module, mock_aos_vni_pool):
```
Comment (roman-verchikov) [inline tests/test_aos_vni_pool.py]: mock_module is unused

File: tests/test_aos_ip_pool.py [hunk 51]
```diff
@@ -0,0 +1,126 @@
+from ansible.compat.tests.mock import patch
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: missing copyright header

File: tests/test_aos_ip_pool.py [hunk 52]
```diff
@@ -0,0 +1,126 @@
...
+@patch('library.aos_ip_pool.AnsibleModule')
+def test_aos_ip_pool_success(mock_module, mock_aos_ip_pool):
+    """
+    aos_ip_pool - test module returned data
+    """
+    mock_session = {"server": "foo",
+                    "token": "bar"}
+
+    mock_aos_ip_pool.return_value = mock_session
+    resp = aos_ip_pool.ip_pool()
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: `ip_pool()` requires `module` argument passed, test seem to be testing mocked instance of ip_pool (in other words, does not test anything)

File: tests/test_aos_ip_pool.py [hunk 53]
```diff
@@ -0,0 +1,126 @@
...
+
+@patch('library.aos_ip_pool.ip_pool')
+@patch('library.aos_ip_pool.AnsibleModule')
+def test_module_args(mock_module,
+                     mock_aos_ip_pool):
+    """
+    aos_ip_pool - test module arguments
+    """
+    aos_ip_pool.main()
+    mock_module.assert_called_with(
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: Not sure what this test does. It seems you already have higher-level tests that cover arguments being passed to playbook, so this can be removed

File: tests/test_aos_ip_pool.py [hunk 54]
```diff
@@ -0,0 +1,126 @@
...
+
+    mock_aos_ip_pool.return_value = mock_session
+    resp = aos_ip_pool.ip_pool()
+    assert resp == mock_session
+
+
+class TestIpPoolValidate(object):
+
+    def setup(self):
+        self.addr_type = 'ipv4'
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: there's no need for `addr_type` to be an instance variable, it could be defined as a global or class variable, and thus would not require the object to maintain state. I'm not sure how and when the `setup()` is called, but it's just an overkill to have it

File: tests/test_aos_ip_pool.py [hunk 55]
```diff
@@ -0,0 +1,126 @@
...
+        test_range = ['192.168.359.0/24']
+        assert aos_ip_pool.validate_subnets(test_range,
+                                            self.addr_type) == ['Invalid subnet: '
+                                                                '192.168.359.0/24']
+
+
+class TestIpv6PoolValidate(object):
+
+    def setup(self):
+        self.addr_type = 'ipv6'
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: same as above

File: tests/test_aos_vni_pool.py [hunk 56]
```diff
@@ -0,0 +1,107 @@
...
+
+
+@patch('library.aos_vni_pool.vni_pool')
+@patch('library.aos_vni_pool.AnsibleModule')
+def test_module_args(mock_module,
+                     mock_aos_vni_pool):
+    """
+    aos_vni_pool - test module arguments
+    """
+    aos_vni_pool.main()
```
Comment (roman-verchikov) [inline tests/test_aos_vni_pool.py]: same as aos_ip_pool.

File: tox.ini [hunk 57]
```diff
@@ -6,12 +6,18 @@ envlist = py37, flake8
 deps =
     -rrequirements.txt
     -rtest-requirements.txt
+passenv =
+    VAULT_PASSWORD
 commands =
-    nosetests -v
+    bash -ec 'if [[ $VAULT_PASSWORD ]]; then echo "$VAULT_PASSWORD" > vault_pass.txt; fi'
+    pytest -q tests/
```
Comment (roman-verchikov) [inline tox.ini]: IMHO it would be better to increase verbosity instead of decreasing it. `pytest -vv` would print the difference between expected and actual values for each failing test, which is very useful for debugging.

**Review by roman-verchikov — CHANGES_REQUESTED** [review]

File: tests/test_aos_ip_pool.py (hunk 53 above)
Comment (that1guy15) [reply tests/test_aos_ip_pool.py]: We have already covered this.
These test insure the args past into Ansible meet the requirements configured and what we expect.

File: tests/test_aos_vni_pool.py (hunk 49 above)
Comment (roman-verchikov) [reply tests/test_aos_vni_pool.py]: unresolved

File: tests/test_aos_vni_pool.py (hunk 50 above)
Comment (roman-verchikov) [reply tests/test_aos_vni_pool.py]: unresolved

File: tests/test_aos_login.py (hunk 47 above)
Comment (roman-verchikov) [reply tests/test_aos_login.py:6]: unresolved

File: tests/test_aos_ip_pool.py [hunk 58]
```diff
@@ -0,0 +1,126 @@
...
+                           'choices': ['ipv4', 'ipv6'],
+                           'default': 'ipv4'}},
+        mutually_exclusive=[('name', 'id')],
+        required_one_of=[('name', 'id')],
+        supports_check_mode=True)
+
+
+@patch('library.aos_ip_pool.ip_pool')
+@patch('library.aos_ip_pool.AnsibleModule')
+def test_aos_ip_pool_success(mock_module, mock_aos_ip_pool):
```
Comment (roman-verchikov) [inline tests/test_aos_ip_pool.py]: I guess it would be better to remove this test, because it's quite confusing and it's not clear what exactly is being tested.

File: tests/test_aos_vni_pool.py [hunk 59]
```diff
@@ -0,0 +1,109 @@
...
+                       'ranges': {'required': False,
+                                  'type': 'list',
+                                  'default': []}},
+        mutually_exclusive=[('name', 'id')],
+        required_one_of=[('name', 'id')], supports_check_mode=True)
+
+
+@patch('library.aos_vni_pool.vni_pool')
+@patch('library.aos_vni_pool.AnsibleModule')
+def test_aos_vni_pool_success(mock_module, mock_aos_vni_pool):
```
Comment (roman-verchikov) [inline tests/test_aos_vni_pool.py]: same here, would be better to remove this test

File: tox.ini [hunk 60]
```diff
@@ -6,12 +6,18 @@ envlist = py37, flake8
 deps =
     -rrequirements.txt
     -rtest-requirements.txt
+passenv =
+    VAULT_PASSWORD
 commands =
-    nosetests -v
+    bash -ec 'if [[ $VAULT_PASSWORD ]]; then echo "$VAULT_PASSWORD" > vault_pass.txt; fi'
+    pytest -q tests/ -vv
```
Comment (roman-verchikov) [inline tox.ini]: remove the `-q` argument from pytest cli call, `-q` and `-vv` are mutually exclusive (-q == "quiet" vs -vv == "very verbose")

File: requirements.txt (hunk 44 above)
Comment (roman-verchikov) [reply requirements.txt:3]: unresolved

**Review by roman-verchikov — APPROVED** [review]
Looks ok with the tox dependency in requirements.txt addressed
